### ISSUE TITLE
Metal backend leads to NaN with take_along_axis

### ISSUE THREAD

**Issue by tdgfrost** [google/jax#17344]
### Description

I have noticed that jnp.take_along_axis() doesn't seem to index correctly when using the Metal backend. 

The following code works with both the Metal backend and standard CPU:

```
import jax.numpy as jnp
import jax.random as random

# jax.config.update('jax_platform_name', 'cpu')
jax.config.update("jax_debug_nans", True)

key = random.PRNGKey(123)
array = random.uniform(key, shape=(3, 4))

idx = jnp.array([[0, 1, 2],[0, 1, 2],[0, 1, 2]])
# idx = jnp.array([[0],[0],[0]])

jnp.take_along_axis(array, idx, axis=1)
```

If you change index to the commented line, it still runs fine on M1 CPU, but the Metal backend gives the following error:

```
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-10-bb9fdf754383>", line 1, in <module>
    jnp.take_along_axis(temp, jnp.array([[0],[1],[2]]), axis=1)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py", line 3999, in take_along_axis
    return lax.gather(arr, gather_indices_arr, dnums, tuple(slice_sizes),
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/lax/slicing.py", line 287, in gather
    return gather_p.bind(
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/core.py", line 380, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/core.py", line 383, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/core.py", line 790, in process_primitive
    return primitive.impl(*tracers, **params)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/dispatch.py", line 144, in apply_primitive
    return compiled_fun(*args)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/dispatch.py", line 227, in <lambda>
    return lambda *args, **kw: compiled(*args, **kw)[0]
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/profiler.py", line 314, in wrapper
    return func(*args, **kwargs)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/interpreters/pxla.py", line 1353, in __call__
    dispatch.check_special(self.name, arrays)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/dispatch.py", line 437, in check_special
    _check_special(name, buf.dtype, buf)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/IQL/lib/python3.10/site-packages/jax/_src/dispatch.py", line 442, in _check_special
    raise FloatingPointError(f"invalid value (nan) encountered in {name}")
FloatingPointError: invalid value (nan) encountered in jit(gather)
```

I think I have isolated the problem to line 3999 in jax/_src/numpy/lax_numpy.py: 

`lax.gather(arr, gather_indices_arr, dnums, tuple(slice_sizes),
                    mode="fill" if mode is None else mode)`

Which leads to line 287 in jax/_src/lax/slicing.py:

`gather_p.bind(
      operand, start_indices, dimension_numbers=dimension_numbers,
      slice_sizes=core.canonicalize_shape(slice_sizes),
      unique_indices=bool(unique_indices),
      indices_are_sorted=bool(indices_are_sorted),
      mode=parsed_mode,
      fill_value=fill_value)`

Which leads to line 1351 in jax/_src/interpreters/pxla.py:

`out_arrays = results.disassemble_into_single_device_arrays()`

where `nan` values start to appear. However, presumably the issue itself occurs a couple of lines before this, at `results = self.xla_executable.execute_sharded(input_bufs)`. I think this may be the same issue with sharding as mentioned in issue  #16287. Any advice on how to address this would be super helpful.

### What jax/jaxlib version are you using?

jax v0.4.11, jaxlib v0.4.10, jax-metal v0.0.3

### Which accelerator(s) are you using?

M1 GPU

### Additional system info

Python 3.10.12, Mac OS 13.3, M1 Pro 10-core / 16GB

**Comment by tdgfrost** [comment]
To add - at the moment I'm getting around this by using lax.gather directly (hopefully correctly), which seems to work fine on both CPU and Metal.

```
from jax import lax

lax.gather(array, 
           jnp.concatenate((jnp.arange(len(idx)).reshape(-1, 1), idx), axis=1),
           lax.GatherDimensionNumbers(
                 offset_dims=tuple([]),
                 collapsed_slice_dims=tuple([0, 1]),
                 start_index_map=tuple([0, 1])), 
           tuple([1, 1]))
```

**Comment by shuhand0** [comment]
The issue is fixed by jax-metal 0.0.5.
